include/picongpu/plugins/xrayScattering/xrayScatteringUtilities.hpp

```diff
@@ -0,0 +1,121 @@
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * PIConGPU is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with PIConGPU.
+ * If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#pragma once
+
+#include "picongpu/simulation_defines.hpp"
+#include <pmacc/math/Vector.hpp>
+#include <pmacc/math/Complex.hpp>
+#include <pmacc/memory/buffers/Buffer.hpp>
+
+
+namespace picongpu
+{
+namespace plugins
+{
+namespace xrayScattering
+{
+    template< typename T >
+    std::vector< T > extractReal(
+        Buffer< pmacc::math::Complex< T >, DIM1 > & complexBuffer )
+    {
+        std::vector< T > realValues;
+        auto size = complexBuffer.getCurrentSize( );
+        auto dataBox = complexBuffer.getDataBox( );
+        realValues.reserve( size );
+        for( uint32_t ii = 0; ii < size;  ii++ )
+        {
+            realValues.push_back( dataBox[ ii ].get_real( ) );
+        }
+        return realValues;
+    }
+
+    template< typename T >
+    std::vector< T > extractImag(
+        Buffer< pmacc::math::Complex< T >, DIM1 > & complexBuffer )
+    {
+        std::vector< T > imagValues;
+        auto size = complexBuffer.getCurrentSize( );
+        auto dataBox = complexBuffer.getDataBox( );
+        imagValues.reserve( size );
+        for( uint32_t ii = 0; ii < size;  ii++ )
+        {
+            imagValues.push_back( dataBox[ ii ].get_imag( ) );
+        }
+        return imagValues;
+    }
+
+    template< typename T >
+    std::vector< T > extractReal(
+        std::vector< pmacc::math::Complex< T > > & complexVec )
+    {
+        std::vector< T > realValues;
+        realValues.reserve( complexVec.size( ) );
+
+        std::transform(
+            std::begin( complexVec ),
+            std::end( complexVec ),
+            std::back_inserter( realValues ),
+            [ ]( const pmacc::math::Complex< T > & data )
+            {
+                return data.get_real( );
+            }
+        );
+        return realValues;
+    }
+
+    template< typename T >
+    std::vector< T > extractImag(
+        std::vector< pmacc::math::Complex< T > > & complexVec )
+    {
+        std::vector< T > imagValues;
+        imagValues.reserve( complexVec.size( ) );
+
+        std::transform(
+            std::begin( complexVec ),
+            std::end( complexVec ),
+            std::back_inserter( imagValues ),
+            [ ]( const pmacc::math::Complex< T > & data )
+            {
+                return data.get_imag( );
+            }
+        );
+        return imagValues;
+    }
+
+    template < typename T >
+    void copyVectorToBuffer(
+        std::vector< T > & vec,
```

Comment:
The vector can become `const`.

Reply:
@sbastrakov Can I make `Buffer` here `const` as well? I'm changing the values... but not the `Buffer` object itself.

Reply:
I think better not, as `vec` is constant input parameter and buffer is output, so at least logically it's not constant. And ideally such code should not compile (but I believe it would in this case).